### DISCUSSION TITLE
feat(Menu): Add `onClose` prop and `@close` event

### DISCUSSION THREAD

**Discussion by phegman** [tailwindlabs/headlessui#1947] [Ideas]
I am building a filter component and I want to be able to reset the options if a user closes the dropdown without selecting a filter option. Currently it doesn't seem like there is a way to detect when the Menu component is closed. 

**Example use case**

https://user-images.githubusercontent.com/4230489/197367787-c45fe7b5-bc16-4835-bf03-860bbfa3dfcd.mov

I would like the dropdown to reset after the user selects `Filter 3` and then closes the dropdown but there is currently no way to detect that the dropdown was closed.

**Comment by gztomas** [comment]
I would like to bump this. It sounds like a pretty important missing feature. This is for example a Linear filter. The search text is cleared every time the menu/combobox is open/closed. There might be a couple of hacky ways to accomplish that with HUI today. Maybe adding a `Transition` and using `afterLeave` which feels wrong if a transition is not wanted. Or using `unmount` and moving the state to some children.

https://user-images.githubusercontent.com/1714020/223802791-d0deaa1b-36e1-4768-a056-1b156e2cca7f.mov


There is a Combobox example in the HUI page showing how `onClose` is needed:

<img width="779" alt="image" src="https://user-images.githubusercontent.com/1714020/223805891-73468009-c636-4d1f-8381-098733a5c01f.png">

**Comment by mikemathewson** [comment]
Save the below code into your own Menu.jsx and in your code replace the Menu from headlessui with this one.

```jsx
import { Menu } from "@headlessui/react";
import { useEffect, useRef } from "react";

const MyMenu = ({ children, onOpen, onClose, ...props }) => {
	return (
		<Menu {...props}>
			{({ open }) => <MenuEventHandler children={children} open={open} onOpen={onOpen} onClose={onClose} />}
		</Menu>
	);
};

const MenuEventHandler = ({ children, open, onOpen, onClose }) => {
	const initialMount = useRef(true);

	useEffect(() => {
		if (!initialMount.current) {
			open ? onOpen && onOpen() : onClose && onClose();
		}
		initialMount.current = false;
	}, [open]);

	return children;
};

export { MyMenu as Menu }
```

**Comment by zanona** [comment]
Please consider implementing this. I saw myself needing this functionality in multiple scenarios.
Found this through #1948, which has been rejected.
